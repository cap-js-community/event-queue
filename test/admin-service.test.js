"use strict";

jest.mock("@opentelemetry/api", () => require("./mocks/openTelemetry"));
const { Logger: mockLogger } = require("./mocks/logger");

const cds = require("@sap/cds");

const project = __dirname + "./../"; // The project's root folder
const { GET } = cds.test(project);

const eventQueue = require("../src");

let loggerMock;

describe("admin-service-test", () => {
  let context, tx;

  beforeEach(async () => {
    context = new cds.EventContext({ user: "testUser", tenant: 123 });
    tx = cds.tx(context);
    jest.clearAllMocks();
    loggerMock = mockLogger();
  });

  afterEach(async () => {
    await tx.rollback();
    jest.clearAllMocks();
  });

  afterAll(() => cds.shutdown);

  it("not allowed if disabled by config", async () => {
    await expect(
      GET("/odata/v4/event-queue/admin/Event", { auth: { username: "yves", password: "" } })
    ).rejects.toMatchSnapshot();
    expect(loggerMock.callsLengths().error).toEqual(0);
  });

  it("metadata snapshot", async () => {
    eventQueue.config.enableAdminService = true;
    const response = await GET("/odata/v4/event-queue/admin/$metadata", { auth: { username: "yves", password: "" } });
    expect(response.data).toMatchSnapshot();
    expect(loggerMock.callsLengths().error).toEqual(0);
  });

  it("read entities", async () => {
    eventQueue.config.enableAdminService = true;
    const response = await GET("/odata/v4/event-queue/admin", { auth: { username: "yves", password: "" } });
    for (const entity of response.data.value) {
      const name = entity.name;
      const metadataResponse = await GET(`/odata/v4/event-queue/admin/${name}`, {
        auth: { username: "yves", password: "" },
      });
      expect(metadataResponse.data).toMatchSnapshot(name);
    }
    expect(loggerMock.callsLengths().error).toEqual(0);
  });
});
