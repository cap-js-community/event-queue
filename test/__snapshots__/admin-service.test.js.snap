// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`admin-service-test metadata snapshot 1`] = `
"<?xml version="1.0" encoding="utf-8"?>
<edmx:Edmx Version="4.0" xmlns:edmx="http://docs.oasis-open.org/odata/ns/edmx">
  <edmx:Reference Uri="https://oasis-tcs.github.io/odata-vocabularies/vocabularies/Org.OData.Capabilities.V1.xml">
    <edmx:Include Alias="Capabilities" Namespace="Org.OData.Capabilities.V1"/>
  </edmx:Reference>
  <edmx:Reference Uri="https://sap.github.io/odata-vocabularies/vocabularies/Common.xml">
    <edmx:Include Alias="Common" Namespace="com.sap.vocabularies.Common.v1"/>
  </edmx:Reference>
  <edmx:Reference Uri="https://oasis-tcs.github.io/odata-vocabularies/vocabularies/Org.OData.Core.V1.xml">
    <edmx:Include Alias="Core" Namespace="Org.OData.Core.V1"/>
  </edmx:Reference>
  <edmx:Reference Uri="https://oasis-tcs.github.io/odata-vocabularies/vocabularies/Org.OData.Validation.V1.xml">
    <edmx:Include Alias="Validation" Namespace="Org.OData.Validation.V1"/>
  </edmx:Reference>
  <edmx:DataServices>
    <Schema Namespace="EventQueueAdminService" xmlns="http://docs.oasis-open.org/odata/ns/edm">
      <Annotation Term="Core.Links">
        <Collection>
          <Record>
            <PropertyValue Property="rel" String="author"/>
            <PropertyValue Property="href" String="https://cap.cloud.sap"/>
          </Record>
        </Collection>
      </Annotation>
      <EntityContainer Name="EntityContainer">
        <EntitySet Name="Event" EntityType="EventQueueAdminService.Event"/>
        <EntitySet Name="Lock" EntityType="EventQueueAdminService.Lock"/>
      </EntityContainer>
      <EntityType Name="Event">
        <Key>
          <PropertyRef Name="ID"/>
        </Key>
        <Property Name="tenant" Type="Edm.String"/>
        <Property Name="landscape" Type="Edm.String"/>
        <Property Name="space" Type="Edm.String"/>
        <Property Name="ID" Type="Edm.Guid" Nullable="false"/>
        <Property Name="type" Type="Edm.String" Nullable="false"/>
        <Property Name="subType" Type="Edm.String" Nullable="false"/>
        <Property Name="referenceEntity" Type="Edm.String"/>
        <Property Name="referenceEntityKey" Type="Edm.Guid"/>
        <Property Name="status" Type="Edm.Int32" Nullable="false" DefaultValue="0"/>
        <Property Name="payload" Type="Edm.String"/>
        <Property Name="attempts" Type="Edm.Int32" Nullable="false" DefaultValue="0"/>
        <Property Name="lastAttemptTimestamp" Type="Edm.DateTimeOffset" Precision="7"/>
        <Property Name="createdAt" Type="Edm.DateTimeOffset" Precision="7"/>
        <Property Name="startAfter" Type="Edm.DateTimeOffset" Precision="7"/>
        <Property Name="context" Type="Edm.String"/>
        <Property Name="error" Type="Edm.String"/>
      </EntityType>
      <EntityType Name="Lock">
        <Key>
          <PropertyRef Name="tenant"/>
          <PropertyRef Name="type"/>
          <PropertyRef Name="subType"/>
        </Key>
        <Property Name="tenant" Type="Edm.String" Nullable="false"/>
        <Property Name="type" Type="Edm.String" Nullable="false"/>
        <Property Name="subType" Type="Edm.String" Nullable="false"/>
        <Property Name="landscape" Type="Edm.String"/>
        <Property Name="space" Type="Edm.String"/>
        <Property Name="ttl" Type="Edm.Int32"/>
        <Property Name="createdAt" Type="Edm.Int32"/>
      </EntityType>
      <Action Name="setStatusAndAttempts" IsBound="true" EntitySetPath="in">
        <Parameter Name="in" Type="EventQueueAdminService.Event"/>
        <Parameter Name="tenant" Type="Edm.String"/>
        <Parameter Name="status" Type="Edm.Int32"/>
        <Parameter Name="attempts" Type="Edm.Int32"/>
        <ReturnType Type="EventQueueAdminService.Event"/>
      </Action>
      <Action Name="releaseLock" IsBound="true">
        <Parameter Name="in" Type="EventQueueAdminService.Lock"/>
        <Parameter Name="tenant" Type="Edm.String"/>
        <Parameter Name="type" Type="Edm.String"/>
        <Parameter Name="subType" Type="Edm.String"/>
        <ReturnType Type="Edm.Boolean"/>
      </Action>
      <Annotations Target="EventQueueAdminService.EntityContainer/Event">
        <Annotation Term="Capabilities.DeleteRestrictions">
          <Record Type="Capabilities.DeleteRestrictionsType">
            <PropertyValue Property="Deletable" Bool="false"/>
          </Record>
        </Annotation>
        <Annotation Term="Capabilities.InsertRestrictions">
          <Record Type="Capabilities.InsertRestrictionsType">
            <PropertyValue Property="Insertable" Bool="false"/>
          </Record>
        </Annotation>
        <Annotation Term="Capabilities.UpdateRestrictions">
          <Record Type="Capabilities.UpdateRestrictionsType">
            <PropertyValue Property="Updatable" Bool="false"/>
          </Record>
        </Annotation>
      </Annotations>
      <Annotations Target="EventQueueAdminService.Event/tenant">
        <Annotation Term="Core.Computed" Bool="true"/>
      </Annotations>
      <Annotations Target="EventQueueAdminService.Event/landscape">
        <Annotation Term="Core.Computed" Bool="true"/>
      </Annotations>
      <Annotations Target="EventQueueAdminService.Event/space">
        <Annotation Term="Core.Computed" Bool="true"/>
      </Annotations>
      <Annotations Target="EventQueueAdminService.Event/ID">
        <Annotation Term="Core.ComputedDefaultValue" Bool="true"/>
      </Annotations>
      <Annotations Target="EventQueueAdminService.Event/status">
        <Annotation Term="Validation.AllowedValues">
          <Collection>
            <Record Type="Validation.AllowedValue">
              <Annotation Term="Core.SymbolicName" String="Open"/>
              <PropertyValue Property="Value" Int="0"/>
            </Record>
            <Record Type="Validation.AllowedValue">
              <Annotation Term="Core.SymbolicName" String="InProgress"/>
              <PropertyValue Property="Value" Int="1"/>
            </Record>
            <Record Type="Validation.AllowedValue">
              <Annotation Term="Core.SymbolicName" String="Done"/>
              <PropertyValue Property="Value" Int="2"/>
            </Record>
            <Record Type="Validation.AllowedValue">
              <Annotation Term="Core.SymbolicName" String="Error"/>
              <PropertyValue Property="Value" Int="3"/>
            </Record>
            <Record Type="Validation.AllowedValue">
              <Annotation Term="Core.SymbolicName" String="Exceeded"/>
              <PropertyValue Property="Value" Int="4"/>
            </Record>
            <Record Type="Validation.AllowedValue">
              <Annotation Term="Core.SymbolicName" String="Suspended"/>
              <PropertyValue Property="Value" Int="5"/>
            </Record>
          </Collection>
        </Annotation>
      </Annotations>
      <Annotations Target="EventQueueAdminService.Event/createdAt">
        <Annotation Term="Core.Computed" Bool="true"/>
      </Annotations>
      <Annotations Target="EventQueueAdminService.setStatusAndAttempts(EventQueueAdminService.Event)/tenant">
        <Annotation Term="Common.FieldControl" EnumMember="Common.FieldControlType/Mandatory"/>
      </Annotations>
      <Annotations Target="EventQueueAdminService.setStatusAndAttempts(EventQueueAdminService.Event)/status">
        <Annotation Term="Validation.AllowedValues">
          <Collection>
            <Record Type="Validation.AllowedValue">
              <Annotation Term="Core.SymbolicName" String="Open"/>
              <PropertyValue Property="Value" Int="0"/>
            </Record>
            <Record Type="Validation.AllowedValue">
              <Annotation Term="Core.SymbolicName" String="InProgress"/>
              <PropertyValue Property="Value" Int="1"/>
            </Record>
            <Record Type="Validation.AllowedValue">
              <Annotation Term="Core.SymbolicName" String="Done"/>
              <PropertyValue Property="Value" Int="2"/>
            </Record>
            <Record Type="Validation.AllowedValue">
              <Annotation Term="Core.SymbolicName" String="Error"/>
              <PropertyValue Property="Value" Int="3"/>
            </Record>
            <Record Type="Validation.AllowedValue">
              <Annotation Term="Core.SymbolicName" String="Exceeded"/>
              <PropertyValue Property="Value" Int="4"/>
            </Record>
            <Record Type="Validation.AllowedValue">
              <Annotation Term="Core.SymbolicName" String="Suspended"/>
              <PropertyValue Property="Value" Int="5"/>
            </Record>
          </Collection>
        </Annotation>
      </Annotations>
      <Annotations Target="EventQueueAdminService.setStatusAndAttempts(EventQueueAdminService.Event)/attempts">
        <Annotation Term="Validation.Minimum" Int="0"/>
        <Annotation Term="Validation.Maximum" Int="100"/>
      </Annotations>
      <Annotations Target="EventQueueAdminService.EntityContainer/Lock">
        <Annotation Term="Capabilities.DeleteRestrictions">
          <Record Type="Capabilities.DeleteRestrictionsType">
            <PropertyValue Property="Deletable" Bool="false"/>
          </Record>
        </Annotation>
        <Annotation Term="Capabilities.InsertRestrictions">
          <Record Type="Capabilities.InsertRestrictionsType">
            <PropertyValue Property="Insertable" Bool="false"/>
          </Record>
        </Annotation>
        <Annotation Term="Capabilities.UpdateRestrictions">
          <Record Type="Capabilities.UpdateRestrictionsType">
            <PropertyValue Property="Updatable" Bool="false"/>
          </Record>
        </Annotation>
      </Annotations>
      <Annotations Target="EventQueueAdminService.releaseLock(EventQueueAdminService.Lock)/tenant">
        <Annotation Term="Common.FieldControl" EnumMember="Common.FieldControlType/Mandatory"/>
      </Annotations>
      <Annotations Target="EventQueueAdminService.releaseLock(EventQueueAdminService.Lock)/type">
        <Annotation Term="Common.FieldControl" EnumMember="Common.FieldControlType/Mandatory"/>
      </Annotations>
      <Annotations Target="EventQueueAdminService.releaseLock(EventQueueAdminService.Lock)/subType">
        <Annotation Term="Common.FieldControl" EnumMember="Common.FieldControlType/Mandatory"/>
      </Annotations>
    </Schema>
  </edmx:DataServices>
</edmx:Edmx>"
`;

exports[`admin-service-test not allowed if disabled by config 1`] = `[AxiosError: 403 - Admin service is disabled by configuration]`;

exports[`admin-service-test read entities: Event 1`] = `
{
  "@odata.context": "$metadata#Event",
  "value": [],
}
`;

exports[`admin-service-test read entities: Lock 1`] = `
{
  "@odata.context": "$metadata#Lock",
  "value": [],
}
`;
